pkg/util/instancemetadata/prodenv_test.go

```diff
@@ -0,0 +1,145 @@
+			environment: map[string]string{
+				"RESOURCEGROUP": "my-rg",
+			},
+			expectErr: true,
+		},
+		{
+			name: "env public cloud",
+			environment: map[string]string{
+				"AZURE_ENVIRONMENT":     azure.PublicCloud.Name,
+				"AZURE_SUBSCRIPTION_ID": "some-sub-guid",
+				"AZURE_TENANT_ID":       "some-tenant-guid",
+				"LOCATION":              "some-region",
+				"RESOURCEGROUP":         "my-resourceGroup",
+			},
+			expectErr: false,
+		},
+		{
+			name: "env fairfax",
+			environment: map[string]string{
+				"AZURE_ENVIRONMENT":     azure.USGovernmentCloud.Name,
+				"AZURE_SUBSCRIPTION_ID": "some-sub-guid",
+				"AZURE_TENANT_ID":       "some-tenant-guid",
+				"LOCATION":              "some-region",
+				"RESOURCEGROUP":         "my-resourceGroup",
+			},
+			expectErr: false,
+		},
+		{
+			name: "env mooncake",
+			environment: map[string]string{
+				"AZURE_ENVIRONMENT":     azure.ChinaCloud.Name,
+				"AZURE_SUBSCRIPTION_ID": "some-sub-guid",
+				"AZURE_TENANT_ID":       "some-tenant-guid",
+				"LOCATION":              "some-region",
+				"RESOURCEGROUP":         "my-resourceGroup",
+			},
+			expectErr: false,
+		},
+		{
+			name: "env blackforest",
+			environment: map[string]string{
+				"AZURE_ENVIRONMENT":     azure.GermanCloud.Name,
+				"AZURE_SUBSCRIPTION_ID": "some-sub-guid",
+				"AZURE_TENANT_ID":       "some-tenant-guid",
+				"LOCATION":              "some-region",
+				"RESOURCEGROUP":         "my-resourceGroup",
+			},
+			expectErr: false,
+		},
+		{
+			name: "env with hostname override",
+			environment: map[string]string{
+				"AZURE_ENVIRONMENT":     azure.PublicCloud.Name,
+				"AZURE_SUBSCRIPTION_ID": "some-sub-guid",
+				"AZURE_TENANT_ID":       "some-tenant-guid",
+				"LOCATION":              "some-region",
+				"RESOURCEGROUP":         "my-resourceGroup",
+				"HOSTNAME_OVERRIDE":     "my.over.ride",
+			},
+			expectErr: false,
+		},
+	}
+
+	for _, test := range tests {
+		testEnv := NewTestEnv(test.environment)
+		p := &prodenv{}
+		err := p.populateInstanceMetadata(testEnv)
+		if test.expectErr != (err != nil) {
+			t.Errorf("%s: expected error %#v got %#v", test.name, test.expectErr, err)
+		} else if !test.expectErr {
+			// verify there are values for all required fields
+			if p.environment == nil {
+				t.Errorf("%s: environment expected, found nil", test.name)
+			}
+			if p.subscriptionID == "" {
+				t.Errorf("%s: subscriptionID expected, found empty string", test.name)
+			}
+			if p.tenantID == "" {
+				t.Errorf("%s: tenantID expected, found empty string", test.name)
+			}
+			if p.location == "" {
+				t.Errorf("%s: location expected, found empty string", test.name)
+			}
+			if p.resourceGroup == "" {
+				t.Errorf("%s: resourceGroup expected, found empty string", test.name)
+			}
+			// hostname is always optional to set env var for but we always expect a value
+			if p.hostname == "" {
+				t.Errorf("%s: hostname expected, found empty string", test.name)
+			}
```

Comment:
Just testing that the value is present is not enough: it doesn't give confidence that the unit under this test works as expected.

`HOSTNAME_OVERRIDE` is a great example: if a value is present - we expec the field to have value from `HOSTNAME_OVERRIDE` env var. If not - we expect value from `os.Hostname()`. 

So if we look at "env with hostname override" test case and at "env public cloud" we expect different values here. But with the current test implemetation we don't know whether or not `populateInstanceMetadata` works as expected.

Reply:
Added expected object with data we want.  I guess some object comparison could work but I went with individual fields so output isn't blob comparison if it is wrong.